keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,376 @@
+  Containers.
+* Secondary: allow actors (users, VPA, StatefulSet, JobController) to decide
+  how to proceed if in-place resource resize is not possible.
+* Secondary: allow users to specify which Pods and Containers can be resized
+  without a restart.
+
+### Non-Goals
+
+The explicit non-goal of this KEP is to avoid controlling full life-cycle of a
+Pod which failed in-place resource resizing. This should be handled by actors
+which initiated the resizing.
+
+Other identified non-goals are:
+* allow to change Pod QoS class without a restart,
+* to change resources of Init Containers without a restart,
+* updating extended resources or any other resource types besides CPU, memory.
+
+## Proposal
+
+### API Changes
+
+PodSpec becomes mutable with regards to Container resources requests and limits.
+Additionally, PodSpec becomes a Pod subresource to allow fine-grained access control.
+
+PodSpec is extended with information about actually allocated Container resources.
+
+Thanks to the above:
+* PodSpec.Container.ResourceRequirements becomes purely a declaration, denoting
+  **desired** state of the Pod resources,
+* PodStatus.ContainerStatus.ResourcesAllocated (new object) shows the resources
+  held by the Pod and its Containers.
+
+In order to determine the state of a Pod resource resize, we add a new PodCondition
+named ResizingPod, which describes the status of the last resize request.
+
+This PodCondition can have the following values:
+* Status: false - Pod resize operation completed,
+  - Reason: (empty) - Initial state, no resize requested since Pod creation,
+  - Reason: Success - Pod and its Containers were successfully resized,
+  - Reason: FailedNodeCapacity - Node does not have room to resize the Pod,
+* Status: true  - Pod is in the process of being resized,
+  - Reason: InProgress - Kubelet is performing Pod resize,
+  - Reason: PreEmpting - Lower priority Pods are being evicted.
+
+#### Container Restart Policy
+
+To provide some fine-grained user control, PodSpec.Container.ResourceRequirements
+is extended with ResizePolicy flag for each resource type (CPU, memory):
+* NoRestart - the default value for CPU; resize Container without restarting it,
+* RestartContainer - the default value for memory; restart the Container in-place
+  to apply new resource values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
```

Comment:
having a separate behavior per resource type may make it harder to define uniform policy controls in the future on top of this feature.  i prefer we keep it simple and having a common behavior for all knobs.

Reply:
We can start with a common policy, and extend it in the future if needed. I believe this approach was chosen so that user could update just memory with restarting, and just CPU without restart (if both CPU and memory are updated, then it would restart the container.)

@kgolab @bskiba Any concerns with this proposed simplification?

Reply:
I do understand your rationale behind here, but shouldn't we agree to have a common policy for all resources initially? I didn't see KEP being updated based on our discussion.

Reply:
never mind, I saw your updates below.

Reply:
This issue has been addressed in the KEP. Resolving.